### PR TITLE
refactor: introduce Fixtures util class

### PR DESCRIPTION
**Pull request by pret-a-porter**
## Changes:

1. Implemented class Fixtures, which provided next functionality: get fixture content by file name, mock filesystem
2. In some tests `loadFixture` replaced by `Fixtures.get`
3. In `editor-config.spec.ts` and `yarn.spec.ts` demonstrated usage of filesystem mock

Since renovate using `fs-extra` `memfs` will be better choice than `mock-fs` for mocking file system API.
If I receive positive feedback about current changes, I will continue refactoring in this way.

## Context:

#10009

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: test/fixtures.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+export class Fixtures {
+  static get(name: string): string {
+    const realFs = jest.requireActual('fs');
+    return realFs.readFileSync(upath.resolve(Fixtures.pathToFixtures, name), {
+      encoding: 'utf-8',
+    });
+  }
+
+  static mock(json: DirectoryJSON, cwd?: string): void {
```
Comment (zharinov) [inline test/fixtures.ts:14]: Maybe, switch argument order would be a good idea. At some point, we may want to extend this class to handle HTTP mocks as well. In this case, value of the first argument may be `/tmp/foo/bar` or `http://example.com/foo/bar`, while the second one defines content mapping for sub-path.

I didn't think a lot about how this API may look like exactly, it's more like just my random thought for now.

Comment (pret-a-porter) [reply test/fixtures.ts:14]: Basically it is just proxy of `vol.fromJSON` from `memfs`, therefore order of arguments also just copied. Also second argument is optional, therefore swap of order of arguments is not possible.

**Comment by pret-a-porter** [comment]
The main concern of current implementation, that no stats about invoked methods like `expect(fs.readFile).toHaveBeenCalledTimes(...)`

File: test/fixtures.ts (hunk 1 above)
Comment (zharinov) [reply test/fixtures.ts:14]: Okay, I see. Though, maybe not impossible, but is more complicated then just argument swap.

File: package.json [hunk 2]
```diff
@@ -136,6 +136,7 @@
     "azure-devops-node-api": "11.0.1",
     "bunyan": "1.8.15",
     "cacache": "15.3.0",
+    "callsite": "1.0.0",
```
Comment (rarkins) [inline package.json:139]: Not a dev dependency?

**Review by zharinov — DISMISSED** [review]
Looks good to me

File: __mocks__/fs.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import { fs } from 'memfs';
+
+module.exports = fs;
```
Comment (viceice) [inline __mocks__/fs.ts:3]: This should be `export = fs`

Do we really need this?

File: lib/datasource/cdnjs/index.spec.ts [hunk 4]
```diff
@@ -1,12 +1,9 @@
 import { getPkgReleases } from '..';
+import { Fixtures } from '../../../test/fixtures';
 import * as httpMock from '../../../test/http-mock';
-import { loadFixture } from '../../../test/util';
 import { EXTERNAL_HOST_ERROR } from '../../constants/error-messages';
 import { CdnJsDatasource } from '.';
 
-const res1 = loadFixture('d3-force.json');
-const res2 = loadFixture('bulma.json');
```
Comment (viceice) [inline lib/datasource/cdnjs/index.spec.ts:8]: I think we should do this wherever possible to allow better garbage collection in tests and reduce memory pressure.

File: lib/workers/branch/auto-replace.spec.ts [hunk 5]
```diff
@@ -1,24 +1,36 @@
...
         ...JSON.parse(JSON.stringify(defaultConfig)),
         manager: 'html',
+        packageFile: 'test',
       };
       reuseExistingBranch = false;
     });
+    afterAll(() => {
+      GlobalConfig.reset();
+    });
+
```
Comment (viceice) [inline lib/workers/branch/auto-replace.spec.ts:33]: I don't think that this is required.

File: lib/workers/branch/auto-replace.spec.ts [hunk 6]
```diff
@@ -154,7 +166,6 @@ describe('workers/branch/auto-replace', () => {
       upgrade.depIndex = 0;
       upgrade.replaceString =
         'image: "1111111111.dkr.ecr.us-east-1.amazonaws.com/my-repository:1"\n\n';
-      upgrade.packageFile = 'k8s/base/defaults.yaml';
```
Comment (viceice) [inline lib/workers/branch/auto-replace.spec.ts:157]: Why this change?

File: test/fixtures.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
+import type { PathLike } from 'fs';
+import callsite from 'callsite';
+import { DirectoryJSON, fs, vol } from 'memfs';
+import upath from 'upath';
+
+export class Fixtures {
+  static get(name: string, fixturesRoot = '.'): string {
+    const realFs = jest.requireActual('fs');
```
Comment (viceice) [inline test/fixtures.ts:8]: Needs a type, currently it's `any`

File: lib/datasource/cdnjs/index.spec.ts (hunk 4 above)
Comment (pret-a-porter) [reply lib/datasource/cdnjs/index.spec.ts:8]: I agree, applied similar changes for more places

File: __mocks__/fs.ts (hunk 3 above)
Comment (pret-a-porter) [reply __mocks__/fs.ts:3]: I have tried, `export = fs;` does not work, also `export default fs;` does not work.
Yeah, we need it, because third part libraries can use `fs` under the hood, but we need to test some functionality based on third party library. `editor-config.spec.ts` for example

File: lib/workers/branch/auto-replace.spec.ts (hunk 5 above)
Comment (pret-a-porter) [reply lib/workers/branch/auto-replace.spec.ts:33]: Why not? It is a global object, it can affect another test

File: lib/workers/branch/auto-replace.spec.ts (hunk 6 above)
Comment (pret-a-porter) [reply lib/workers/branch/auto-replace.spec.ts]: I have returned this, `outputFile` function was mocked a bit incorrectly

**Review by zharinov — DISMISSED** [review]

File: lib/workers/branch/auto-replace.spec.ts (hunk 5 above)
Comment (viceice) [reply lib/workers/branch/auto-replace.spec.ts:33]: Nope, it can only affect tests in this file. Jest will isolated each file with fresh modules.
You only need to revert build-in object changes, like our `nock` tools does.

So please remove this.

File: test/fixtures.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+export class Fixtures {
+  static get(name: string, fixturesRoot = '.'): string {
+    const realFs: IFs = jest.requireActual('fs');
+    return realFs.readFileSync(
+      upath.resolve(Fixtures.getPathToFixtures(fixturesRoot), name),
+      {
+        encoding: 'utf-8',
+      }
+    ) as string;
```
Comment (viceice) [inline test/fixtures.ts:14]:
```suggestion
    );
```
Is this required for type safety?

Comment (pret-a-porter) [reply test/fixtures.ts:14]: Yeah, memfs does not have proper overload type for case, when read file function invoked with encoding 'utf-8'

Comment (viceice) [reply test/fixtures.ts:14]: You should not use memfs type for real FS. Use the original type

File: lib/manager/npm/post-update/yarn.spec.ts [hunk 9]
```diff
@@ -242,85 +246,62 @@ describe('manager/npm/post-update/yarn', () => {
...
+        },
+        '/tmp/renovate'
+      );
+      const { offlineMirror, yarnPath } = await yarnHelper.checkYarnrc(
         '/tmp/renovate'
       );
       expect(offlineMirror).toBeFalse();
       expect(yarnPath).toBeNull();
-      expect(yarnrcContents).not.toContain('yarn-path');
+      expect(Fixtures.toJSON('/tmp/renovate/.yarnrc')).not.toContain(
```
Comment (zharinov) [inline lib/manager/npm/post-update/yarn.spec.ts:302]: Not sure this assertion works as expected. If I'm correct, previously it was testing string to contain `yarn-path` substring. Now, return value for `Fixtures.toJSON('/tmp/renovate/.yarnrc')` is `{ '/tmp/renovate/.yarnrc': '\n' }`, and `toContain` doesn't work well for objects.

**Review by zharinov — CHANGES_REQUESTED** [review]

File: test/fixtures.ts [hunk 10]
```diff
@@ -0,0 +1,77 @@
+import type fs from 'fs';
+import type { PathLike } from 'fs';
+import callsite from 'callsite';
+import { DirectoryJSON, fs as memfs, vol } from 'memfs';
+import upath from 'upath';
+
+export class Fixtures {
+  static get(name: string, fixturesRoot = '.'): string {
```
Comment (viceice) [inline test/fixtures.ts:8]: Please add js-doc comments to functions, so everybody knows what they are doing.

File: lib/manager/npm/post-update/yarn.spec.ts (hunk 9 above)
Comment (viceice) [reply lib/manager/npm/post-update/yarn.spec.ts:302]: Yeah, and i don't really know what the `toJSON` function is doing. So see comments below

File: lib/manager/npm/post-update/yarn.spec.ts [hunk 11]
```diff
@@ -242,85 +246,62 @@ describe('manager/npm/post-update/yarn', () => {
...
+    Fixtures.mock({});
     const execSnapshots = mockExecAll(exec, {
       stdout: '1.9.4',
       stderr: 'some-error',
     });
-    fs.readFile.mockResolvedValueOnce(null).mockRejectedValueOnce(() => {
-      throw new Error('not-found');
-    });
     const res = await yarnHelper.generateLockFile('some-dir', {});
-    expect(fs.readFile).toHaveBeenCalledTimes(2);
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.spec.ts:253]: whay we can't have that check?

File: test/fixtures.ts (hunk 10 above)
Comment (pret-a-porter) [reply test/fixtures.ts:17]: Added, let me know if need more information

File: lib/manager/npm/post-update/yarn.spec.ts (hunk 9 above)
Comment (pret-a-porter) [reply lib/manager/npm/post-update/yarn.spec.ts:302]: Changed to `expect(Fixtures.toJSON()['/tmp/renovate/.yarnrc']).toBe('\n');`

File: lib/manager/npm/post-update/yarn.spec.ts (hunk 11 above)
Comment (pret-a-porter) [reply lib/manager/npm/post-update/yarn.spec.ts:253]: Returned, previous implementation did not allowed it

File: lib/manager/npm/post-update/yarn.spec.ts [hunk 12]
```diff
@@ -201,7 +209,7 @@ describe('manager/npm/post-update/yarn', () => {
       ]);
       expect(fs.readFile).toHaveBeenCalledTimes(expectedFsCalls);
       expect(fs.remove).toHaveBeenCalledTimes(1);
-      expect(res.lockFile).toBe('package-lock-contents');
+      expect(res.lockFile).toBeUndefined();
```
Comment (zharinov) [inline lib/manager/npm/post-update/yarn.spec.ts:212]: Why this field became undefined?

Comment (pret-a-porter) [reply lib/manager/npm/post-update/yarn.spec.ts:213]: image

When `upgrades` param of function `generateLockFile` contains `isLockFileMaintenance` we should remove lock file. Therefore previous implementation of unit test was not fully correct.

Comment (pret-a-porter) [reply lib/manager/npm/post-update/yarn.spec.ts:213]: I have added `yarn.lock` file content to Fixture of this unit test to be sure it removes

**Review by zharinov — DISMISSED** [review]
@rarkins @viceice Don't we want to make these changes incremental? How about merging changes @pret-a-porter already have done, and proceed further with smaller steps?

**Comment by rarkins** [comment]
I'm ok to merge this as-is if @viceice agrees

File: test/fixtures.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+  }
+
+  private static getPathToFixtures(fixturesRoot = '.'): string {
+    const stack = callsite();
+    const callerDir = upath.dirname(stack[2].getFileName());
+    return upath.resolve(callerDir, fixturesRoot, '__fixtures__');
+  }
+}
+
+function readFile(fileName: string, encoding?: string): Promise<unknown> {
```
Comment (viceice) [inline test/fixtures.ts:77]: Wouldn't return type be always string, as you always pass encoding?

File: test/fixtures.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  return memfs.promises.readFile(fileName, encoding ?? 'utf8');
+}
+
+export async function outputFile(file: string, data: any): Promise<void> {
```
Comment (viceice) [inline test/fixtures.ts:81]:
```suggestion
export async function outputFile(file: string, data: unknown): Promise<void> {
```

Do not use any if possible

File: test/fixtures.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+ */
+export class Fixtures {
+  /**
+   * Returns content from fixture file from __fixtures__ folder
+   * @param name name of the fixture file
+   * @param fixturesRoot is an optional, by default is current folder
+   * @returns
+   */
+  static get(name: string, fixturesRoot = '.'): string {
+    const realFs = jest.requireActual<typeof fs>('fs');
```
Comment (viceice) [inline test/fixtures.ts:18]: Shouldn't we move this to module scope?

File: test/fixtures.ts [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+import upath from 'upath';
+
+/**
+ * Class to work with in-memory file-system
+ */
+export class Fixtures {
+  /**
+   * Returns content from fixture file from __fixtures__ folder
+   * @param name name of the fixture file
+   * @param fixturesRoot is an optional, by default is current folder
```
Comment (viceice) [inline test/fixtures.ts:14]:
```suggestion
   * @param fixturesRoot is optional, by default is current test file folder
```

File: test/fixtures.ts (hunk 13 above)
Comment (pret-a-porter) [reply test/fixtures.ts:77]: Yes, right it will be. `memfs.promises.readFile` returns `Promise<string | Buffer>` that is the reason why I am using `unknown` here. Basically I can cast type by `as`.

Comment (viceice) [reply test/fixtures.ts:78]: Please declare like this
https://github.com/renovatebot/renovate/blob/e14af43880d4e7e05fba19b83a87261390d69c68/lib/util/fs/index.ts#L25-L45

File: test/fixtures.ts [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+const realFs = jest.requireActual<typeof fs>('fs');
+
+/**
+ * Class to work with in-memory file-system
+ */
+export class Fixtures {
+  /**
+   * Returns content from fixture file from __fixtures__ folder
+   * @param name name of the fixture file
+   * @param fixturesRoot is an optional, by default is current test folder
```
Comment (viceice) [inline test/fixtures.ts:16]:
```suggestion
   * @param fixturesRoot is an optional string, by default is current test folder
```

**Review by viceice — COMMENTED** [review]
@HonkingGoose can you please review the jsdoc in `test/fixtures.ts`?

File: test/fixtures.ts (hunk 17 above)
Comment (HonkingGoose) [reply test/fixtures.ts:16]:
```suggestion
   * @param {string} [fixturesRoot] - Where to find the fixtures, use the current test folder by default
```

I think this is the correct JSDoc syntax, but let me know if I'm doing it wrong!

The idea here is:
- Use `{string}` to say what type the input should be
- Wrap the paramater name in `[]` to mark it _optional_
- Use a hyphen `-` to mark the start of the `description` text

Relevant quotes from JSDocs docs [^docs]:

> **Name, type, and description, with a hyphen before the description**
> 
> ```js
> /**
>  * @param {string} somebody - Somebody's name.
>  */
> function sayHello(somebody) {
>     alert('Hello ' + somebody);
> }
> ```

> **An optional parameter (using JSDoc syntax)**
>
> 
> ```js
> /**
>  * @param {string} [somebody] - Somebody's name.
>  */
> function sayHello(somebody) {
>     if (!somebody) {
>         somebody = 'John Doe';
>     }
>     alert('Hello ' + somebody);
> }
> ```

[^docs]: https://jsdoc.app/tags-param.html

**Review by HonkingGoose — COMMENTED** [review]
First comment, let's see what we think about this first, before I start suggesting more changes. 😉

File: test/fixtures.ts (hunk 17 above)
Comment (viceice) [reply test/fixtures.ts:16]: The type is not required for typescript, as it's inferred from function declaration.

Comment (HonkingGoose) [reply test/fixtures.ts:16]:
```suggestion
   * @param [fixturesRoot] - Where to find the fixtures, uses the current test folder by default
```

Removed `{string}` from my suggestion, as we don't need it. 😉 

Will the rest of this suggestion work okay? Or is there something wrong with that as well? 🙈

Comment (viceice) [reply test/fixtures.ts:16]: Fine bye me

**Comment by rarkins** [comment]
Needs deconflicting

**Review by zharinov — DISMISSED** [review]

File: test/fixtures.ts [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+  private static getPathToFixtures(fixturesRoot = '.'): string {
+    const stack = callsite();
+    const callerDir = upath.dirname(stack[2].getFileName());
+    return upath.resolve(callerDir, fixturesRoot, '__fixtures__');
+  }
+}
+
+function readFile(fileName: string): Promise<Buffer>;
+function readFile(fileName: string, encoding: 'utf8'): Promise<string>;
+function readFile(
```
Comment (viceice) [inline test/fixtures.ts:80]: I'm a little bit confused, but where is this function used when not exported? 🤔

Comment (pret-a-porter) [reply test/fixtures.ts:80]: It uses to mock `fs-extra` properly:

```
static fsExtra(): any {
    return {
      ...memfs,
      pathExists: jest.fn().mockImplementation(pathExists),
      remove: jest.fn().mockImplementation(memfs.promises.rm),
      readFile: jest.fn().mockImplementation(readFile),
      writeFile: jest.fn().mockImplementation(memfs.promises.writeFile),
      outputFile: jest.fn().mockImplementation(outputFile),
    };
  }
```

File: test/fixtures.ts [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+    vol.reset();
+  }
+
+  // Temporary solution, when all tests will be rewritten to Fixtures mocks can be moved into __mocks__ folder
+  static fsExtra(): any {
+    return {
+      ...memfs,
+      pathExists: jest.fn().mockImplementation(pathExists),
+      remove: jest.fn().mockImplementation(memfs.promises.rm),
+      readFile: jest.fn().mockImplementation(readFile),
```
Comment (viceice) [inline test/fixtures.ts:65]:
```suggestion
      readFile: jest.fn().mockImplementation(memfs.promise.readFile),
```

So we don't need the wrapper function?

Comment (pret-a-porter) [reply test/fixtures.ts]: You are absolutely right! To be honest, I did not remember why I Introduced function with default encoding. Changed

**Review by viceice — APPROVED** [review]

**Comment by zharinov** [comment]
Great 🎉🎉🎉
